framework/include/utils/SymmetricRankTwoTensor.h

```diff
@@ -0,0 +1,640 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "Moose.h"
+#include "ADRankTwoTensorForward.h"
+#include "ADSymmetricRankTwoTensorForward.h"
+#include "ADSymmetricRankFourTensorForward.h"
+#include "MooseUtils.h"
+
+#include "libmesh/libmesh.h"
+#include "libmesh/tensor_value.h"
+
+#include "metaphysicl/raw_type.h"
+
+#include <petscsys.h>
+#include <vector>
+
+// Forward declarations
+class MooseEnum;
+template <typename T>
+class MooseArray;
+typedef MooseArray<Real> VariableValue;
+template <typename>
+class ColumnMajorMatrixTempl;
+namespace libMesh
+{
+template <typename>
+class TypeVector;
+template <typename>
+class TypeTensor;
+template <typename>
+class TensorValue;
+}
+
+namespace MathUtils
+{
+template <typename T>
+void mooseSetToZero(T & v);
+
+/**
+ * Helper function template specialization to set an object to zero.
+ * Needed by DerivativeMaterialInterface
+ */
+template <>
+void mooseSetToZero<SymmetricRankTwoTensor>(SymmetricRankTwoTensor & v);
+
+/**
+ * Helper function template specialization to set an object to zero.
+ * Needed by DerivativeMaterialInterface
+ */
+template <>
+void mooseSetToZero<ADSymmetricRankTwoTensor>(ADSymmetricRankTwoTensor & v);
+}
+
+/**
+ * SymmetricRankTwoTensorTempl is designed to handle the Stress or Strain Tensor for
+ * an anisotropic material. It is designed to reduce the redundancies of the
+ * Complete tensor classes for regular mechanics problems and to enable use of the
+ * Mandel notation.
+ */
+template <typename T>
+class SymmetricRankTwoTensorTempl
+{
+public:
+  ///@{ tensor dimension and Mandel vector length
+  static constexpr unsigned int Ndim = 3;
+  static constexpr unsigned int N = 2 * Ndim;
+  ///@}
+
+  // Select initialization
+  enum InitMethod
+  {
+    initNone,
+    initIdentity
+  };
+
+  /// Default constructor; fills to zero
+  SymmetricRankTwoTensorTempl();
+
+  /// Select specific initialization pattern
+  SymmetricRankTwoTensorTempl(const InitMethod);
+
+  /**
+   * To fill up the 6 entries in the 2nd-order tensor, fillFromInputVector
+   * is called with one of the following fill_methods.
+   * See the fill*FromInputVector functions for more details
+   */
+  enum FillMethod
+  {
+    autodetect = 0,
+    isotropic1 = 1,
+    diagonal3 = 3,
+    symmetric6 = 6
+  };
+
+  /**
+   * Constructor that takes in 3 vectors and uses them to create rows
+   * _vals[0][i] = row1(i), _vals[1][i] = row2(i), _vals[2][i] = row3(i)
+   */
+  SymmetricRankTwoTensorTempl(const TypeVector<T> & row1,
+                              const TypeVector<T> & row2,
+                              const TypeVector<T> & row3);
```

Comment:
shouldn't have added it. I want to deprecate this in R2T because it is ambiguous. I replaced it with the named constructors `initializeFromRows` and `initializeFromColumns`